elastic/elastic-agent
# status identifies failing component, fleet gateway may report degraded, liveness endpoint added

## What does this PR do?

Extend the status reporter to include status update time, and the message
from the failing component or app.

Allow the fleet-gateway (component) to report a degraded state if two checkins
fail.

Add /liveness route to metrics server. This route will report the status. A 200 is
returned if the agent is healthy, and a 503 is returned if it is degraded or failed,
the response body is a json object with additional details. Status may not
propagate to fleet-server in the case where degradation is caused by checkin
fails. It may also lead to the server reporting degraded for up to 30s (fleet-server
polling time) when the agent is able to successfully connect.

## Why is it important?

Allowing the agent to report a degraded state when it cannot contact fleet is
useful as it can be used to better debug connectivity issues. Exposing this
information on a liveness endpoint enables platforms (such as k8s) to poll it
and manage the lifecycle of the agent.

## Checklist

- ~~I have made corresponding change to the default configuration files~~

## How to test this PR locally

- Build the agent locally
- Run the elastic-stack with `elastic-package stack up`
- ensure that the agent + fleet server appears in the fleet ui
- stop the fleet-server container
- wait a minute then ensure that elastic-agent container reports a degraded status locally
- `curl` the liveness endpoint and ensure a 503 is returned

## Related issues

- Relates #390
